addons/point_of_sale/models/pos_session.py

```diff
@@ -2099,6 +2108,117 @@ def get_pos_ui_product_pricelist_item_by_product(self, product_tmpl_ids, product
+
+    def _process_order_process_fail(self, order: dict, exception: Exception, uid: int = False):
+        current_pos_order_data_hash = hash(PoSOrderData.from_data(order['data']))
+        self._get_unprocessed_pos_order_scheduled_activity(
+            order['data']['name'],
+            current_pos_order_data_hash,
+            uid or self.env.user.id,
+            create=True,
+        )
+        self._capture_order_data(order, exception, current_pos_order_data_hash)
+
+    def _shorten_pos_order_data_hash(self, pos_order_data_hash: int):
+        # The bigger the value, the less likely a hash collision will occur
+        # but a bigger value also means a longer name in the attachment
+        return str(pos_order_data_hash)[:6]
+
+    def _get_unprocessed_pos_order_scheduled_activity(self, order_ref: str, pos_order_data_hash: int, assigned_user_id: Optional[int] = None, create: bool = True):
+        if create:
+            assert assigned_user_id, "if in creation mode, user is is needed"
+        pos_module_name = 'point_of_sale'
+        order_ref = order_ref.replace(' ', '_')
+        activity_xid_name = f"activity_pos_unprocessed_{order_ref}_{pos_order_data_hash}"
+        scheduled_activity = self.env.ref(f"{pos_module_name}.{activity_xid_name}", raise_if_not_found=False)
+        if create and not scheduled_activity:
+            scheduled_activity = self.activity_schedule(
+                act_type_xmlid='mail.mail_activity_data_warning',
+                summary=_("PoS order %s can not be processed", order_ref),
+                note=_("The Point of Sale order with the following reference <code>%s #%s</code> "
+                       "was received by the Odoo server, but the order processing phase failed.<br/>"
+                       "The datas received from the point of sale has been saved in the attachments.<br/>"
+                       "Please contact your support service to assist you on restoring it",
+                       order_ref, self._shorten_pos_order_data_hash(pos_order_data_hash)),
+                user_id=assigned_user_id,
+            )
+            # and set it an XID
+            self.env['ir.model.data'].create({
+                'name': activity_xid_name,
+                'module': pos_module_name,
+                'model': scheduled_activity._name,
+                'res_id': scheduled_activity.id,
+            })
+        return scheduled_activity
+
+    def _capture_order_data_attachment_name(self, order_ref: str, pos_order_data_hash: int):
+        return f"pos_order_save_{order_ref}_{self._shorten_pos_order_data_hash(pos_order_data_hash)}.json"
+
+    def _get_captured_order_attachment(self, pos_order_ref: str, pos_order_data_hash: int):
+        return self.env['ir.attachment'].search([
+            ['name', '=', self._capture_order_data_attachment_name(pos_order_ref, pos_order_data_hash)],
```

Comment:
add res_model/res_id?

Reply:
@mart-e 
I used to use them in the domain.
But there is one case where the res_id is harder to determine as if we close a PoS session with left-over orders it create a new PoS session (called rescue session) and probably other strange case that we did not thought of.
We can hardcode the res_model to "pos.session" but we thought that with the very strange name of the attachment, searching on the name only was enough.